### ISSUE TITLE
Visibility Range "Fade Self" option hides shadows unexpectedly

### ISSUE THREAD

**Issue by eswartz** [godotengine/godot#91671]
### Tested versions

- Reproducible in 4.3 dev 5 and 7cdad333114e6765351ed0facb48db228ef29b7b

### System information

Godot v4.3.dev (32339a8c3) - Debian GNU/Linux trixie/sid trixie - X11 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 3080 Ti (nvidia) - AMD Ryzen 7 5800X 8-Core Processor (16 Threads)

### Issue description

When using this combination:

* GeometryInstance3D > Visibility Range established
* Fade Mode set to Self
* Add a mesh instance with a material that enables alpha blending
* Enable Cast Shadows on meshes
* Add a light with shadows enabled

-- then no shadows are ever rendered. 

I assume the intention is to have shadows disappear when the object is actually invisible, as it does for an opaque material under the same circumstances.

### Steps to reproduce

See MRP. Move camera back and forth. 

Close up, one shadow missing on "GeoDistanceFadeSelf + Blend":

![screenshot-20240507-125405](https://github.com/godotengine/godot/assets/875407/b2f051ae-eccd-436b-97ad-32b520c25b1d)

At a distance, all shadows gone as expected:

![screenshot-20240507-125415](https://github.com/godotengine/godot/assets/875407/b03dce30-18ab-4e5e-a883-a0644858f342)

### Minimal reproduction project (MRP)

[fade-shadow-test.zip](https://github.com/godotengine/godot/files/15239133/fade-shadow-test.zip)

**Comment by Calinou** [comment]
Materials with alpha transparency never cast shadows. if you want them to cast shadows, you need to duplicate the MeshInstance, give the duplicate an opaque material and set its shadow casting mode to **Shadows Only**.

GeometryInstance3D transparency allows materials that are made transparent this way to cast shadows, but only if the material is originally opaque (or alpha scissor/alpha hash).

**Comment by eswartz** [comment]
I should clarify, the "blend" objects in the project are using the Depth Pre-Pass model, not actual alpha blending. I'll clarify the title.

**Comment by eswartz** [comment]
According to material.cpp:788, using pre-pass should add this `render_mode`:
```

	if (transparency == TRANSPARENCY_ALPHA_DEPTH_PRE_PASS) {
		code += ", depth_prepass_alpha";
	}
```
which should then trigger `casts_shadows` per `uses_depth_prepass_alpha` in the check:
```

bool SceneShaderForwardClustered::ShaderData::casts_shadows() const {
	bool has_read_screen_alpha = uses_screen_texture || uses_depth_texture || uses_normal_texture;
	bool has_base_alpha = (uses_alpha && (!uses_alpha_clip || uses_alpha_antialiasing)) || has_read_screen_alpha;
	bool has_alpha = has_base_alpha || uses_blend_alpha;

	return !has_alpha || (uses_depth_prepass_alpha && !(depth_draw == DEPTH_DRAW_DISABLED || depth_test == DEPTH_TEST_DISABLED));
}
```
The material I'm using seems to pass this test (depth draw is OPAQUE and "no depth test" is not enabled).

![image](https://github.com/godotengine/godot/assets/875407/4bf45a6d-0705-4438-9e77-c79c8c2d420f)

Am I understanding correctly? It seems this should still be considered to cast a shadow.

**Comment by Calinou** [comment]
> Am I understanding correctly? It seems this should still be considered to cast a shadow.

Yes, that makes sense. I recall this bug being reported elsewhere already, but I can't find an issue right now. Perhaps https://github.com/godotengine/godot/issues/73158 is related.
